src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp

```diff
@@ -0,0 +1,72 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===------------- ONNXQuantizeLinearOp.cpp - ONNXQuantizeLinearOp --------------===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file lowers ONNXQuantizeLinearOp operator to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "src/Conversion/ONNXToTOSA/DialectBuilder.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "src/Dialect/ONNX/ONNXOps/ShapeHelper.hpp"
+#include "torch-mlir/Conversion/TorchToTosa/TosaLegalizeUtils.h"
+#include "llvm/ADT/ArrayRef.h"
+#include <src/Dialect/Mlir/IndexExpr.hpp>
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+class ONNXQuantizeLinearOpLoweringToTOSA : public ConversionPattern {
+public:
+  ONNXQuantizeLinearOpLoweringToTOSA(MLIRContext *ctx)
+      : ConversionPattern(ONNXQuantizeLinearOp::getOperationName(), 1, ctx) {}
+  using OpAdaptor = typename ONNXQuantizeLinearOp::Adaptor;
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    // Quantization formula is ((x / y_scale) + y_zero_point)
+    // TODO : Axis and Saturate attributes
+    TosaBuilder tosaBuilder(rewriter, op->getLoc());
+    OpAdaptor adaptor(operands, op->getAttrDictionary());
+    auto qLinearOp = llvm::cast<ONNXQuantizeLinearOp>(op);
+    //mlir::Type resultType =
+    //    getTypeConverter()->convertType(qLinearOp.getResult().getType());
+
+    mlir::Value x = qLinearOp.x();
+    auto y_scale = qLinearOp.y_scale();
+    mlir::Value y_zero_point = qLinearOp.y_zero_point();
+    auto loc = op->getLoc();
+
+    // Replace the division by a reciprocal followed by a mul
+    auto recOp = tosa::CreateOpAndInfer<mlir::tosa::ReciprocalOp>(rewriter, loc, x.getType(), x).getResult();
+    auto mulOp = tosa::CreateOpAndInfer<mlir::tosa::MulOp>(rewriter, loc, x.getType(), recOp, y_scale, 0).getResult();
```

Comment:
An op to round is missing here. It is in the description in https://github.com/onnx/onnx/blob/main/docs/Operators.md#QuantizeLinear

Reply:
Yes the documentation isn't great. I think I misunderstood the rounding for a clamping (which is done by the cast op)

Reply:
Now that I look at it a bit more, it seems that tosa only has `floor` and `ceil` as rounding operator, and nothing to do a more fine-grained rounding, which sounds wrong in all cases. Do you know if there have been discussions around this topic already?

Reply:
Looks like Matthias had the same questions https://discuss.mlplatform.org/t/tosa-op-with-round-half-to-even-lowering-pytorchs-quantize-per-tensor/165/13

Reply:
I read it in a different way (I am not 100% sure though), but let's discuss this tomorrow in the office

Reply:
Discussed offline, cast seems to be doing the rounding properly